pslab/bus/spi.py

```diff
@@ -0,0 +1,505 @@
+            Secondary prescaler (0,1,..7) -> (8:1,7:1,..1:1).
+        CKE : {0, 1}
+            SPIx Clock Edge Select bit. Serial output data changes on transition
+            {0: from Idle clock state to active clock state,
+             1: from active clock state to Idle clock state}.
+        CKP : {0, 1}
+            Clock Polarity Select bit.
+            Idle state for clock is a {0: low, 1: high} level.
+        SMP : {0, 1}
+            Input data is sampled at the {0: end, 1: middle} of data output time.
+
+        Raises
+        ------
+        ValueError
+            If any one of arguments is not in its shown range.
+        """
+        self._set_parameters(primary_prescaler, secondary_prescaler, CKE, CKP, SMP)
+
+
+class SPISlave(SPIPrimitive):
+    """SPI slave device.
+
+    Parameters
+    ----------
+    device : :class:`SerialHandler`, optional
+        Serial connection to PSLab device. If not provided, a new one will be
+        created.
+    """
+
+    def __init__(self, device: SerialHandler = None):
+        super().__init__(device)
+
+    def transfer8(self, data: int) -> int:
+        """Send 8-bit data over SPI and receive 8-bit data from SPI simultaneously.
+
+        Parameters
+        ----------
+        data : int
+            Data to transmit.
+
+        Returns
+        -------
+        data_in : int
+            Data returned by slave device.
+        """
+        self._start()
+        data_in = self._transfer(data, 8)
+        self._stop()
+        return data_in
+
+    def transfer16(self, data: int) -> int:
+        """Send 16-bit data over SPI and receive 16-bit data from SPI simultaneously.
+
+        Parameters
+        ----------
+        data : int
+            Data to transmit.
+
+        Returns
+        -------
+        data_in : int
+            Data returned by slave device.
+        """
+        self._start()
+        data_in = self._transfer(data, 16)
+        self._stop()
+        return data_in
+
+    def transfer8_bulk(self, data: List[int]) -> List[int]:
+        """Transfer 8-bit data array over SPI.
+
+        Parameters
+        ----------
+        data : list of int
+            List of 8-bit data to transmit.
+
+        Returns
+        -------
+        data_in : list of int
+            List of 8-bit data returned by slave device.
+        """
+        self._start()
+        data_in = self._transfer_bulk(data, 8)
+        self._stop()
+        return data_in
+
+    def transfer16_bulk(self, data: List[int]) -> List[int]:
+        """Transfer 16-bit data array over SPI.
+
+        Parameters
+        ----------
+        data : list of int
+            List of 16-bit data to transmit.
+
+        Returns
+        -------
+        data_in : list of int
+            List of 16-bit data returned by slave device.
+        """
+        self._start()
+        data_in = self._transfer_bulk(data, 16)
+        self._stop()
+        return data_in
+
+    def read8(self) -> int:
```

Comment:
Should we provide `write*` methods as well, for symmetry with these `read*` methods?

Reply:
`write*` will be simply `transfer*` without returning what we read.
For sake of symmetry, we can add these methods :+1:.